### PR TITLE
Fix webcmdlets so that an empty `Get` does not include a `content-length` header

### LINKED ISSUE #15186: Invoke-WebRequest/Invoke-RestMethod: PR #10034 violates RFC7230 and therefore breaks HTTP Desync Mitigation modes
## Background

PR #10034 that fixes issue #9473 and has been introduced in `7.0.0-preview.2`.
This sets a `String.Empty` body for GET requests in `Invoke-WebRequest` and `Invoke-RestMethod` which usually do not have a body.

Unfortunately this also leads .NET to add a `Content-Length: 0` header because of the `String.Empty` HTTP request body.

And this breaks RFC7230, see the official RFC7230 Section 3.3.2: https://tools.ietf.org/html/rfc7230#section-3.3.2 
```
   A user agent SHOULD NOT send a
   Content-Length header field when the request message does not contain
   a payload body and the method semantics do not anticipate such a
   body.
```
Why is this a problem?
* To prevent HTTP Request Smuggling attacks Cloud Service Providers such as AWS and vendors of Application Load Balancers have implemented so-called "HTTP Desync Mitigation modes".
* On AWS the "Strictest Mode" for example blocks all requests that are not RFC7230 conform.
** See https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#desync-mitigation-mode for explanation
* Here is a list of classification reasons that are not RFC7230 conform: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-access-logs.html#classification-reasons
** `GetHeadZeroContentLength` - There is a `Content-Length` header with a value of 0 for a GET or HEAD request.
** This classifies the request as `Acceptable` which is being blocked in "Strictest Mode"
* You cannot use `Invoke-WebRequest` or `Invoke-RestMethod` on AWS hosted websites where this setting has been set
* This will result in a HTTP 400 - Bad Request
* This might also affect other Cloud Service Providers like Azure or Google Cloud Platform, but I cannot confirm that, as I'm only using AWS.

## Steps to reproduce

* On an AWS Application Load Balancer set the HTTP Desync Mitigation Mode to "Strictest Mode"
* Then try a simple `Invoke-WebRequest` on `PowerShell 7.0.0-preview.2` or above:
```powershell
MY C:\> Invoke-WebRequest -Uri "https://mywebsite.example.com/"
Invoke-WebRequest:
400 Bad Request
400 Bad Request
```

## Expected behavior

* Do not set a `Content-Length: 0` header as it is not RFC7230 conform if you do not have a HTTP Request body

## Actual behavior

* PowerShell 7 always sets the HTTP header `Content-Length: 0` for GET/HEAD requests without HTTP Request body and therefore violates RFC7230

## Environment data

* This is reproducible on AWS Lambda with `PowerShell 7.0.0` on Linux as well as on Windows 10 with `PowerShell 7.1.0`

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

This change https://github.com/PowerShell/PowerShell/pull/10034 sets an empty string for the content which results in a `content-length` header with value 0.  This was to fix a case where a web service doesn't accept explicit empty content for a `put` call.  However, this broke a case with a `get` call where normally you don't send content and the server side is failing the request when we specify a `content-length` header with value 0.

So the change here is to only add the `content-length` header for empty content if it's not a `get` method call.

I was running the tests locally and the numerous progress bars being rendered created a different problem (hitting a debug assert), so to avoid that I suppressed all progress in the tests as it's not really used for the tests.  Existing `get` test was modified to check that this header isn't being added.

Also had to update dotnet links so that static analysis passes as their old links no longer work.

## PR Context

If you use:

```powershell
invoke-restmethod https://api.npmjs.org/downloads/range/last-month/@microsoft/mgt
```

This fails with the server reporting a 500 (internal server error), but works with Windows PowerShell 5.1.

Fix https://github.com/PowerShell/PowerShell/issues/15186

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1281]: You could put this check for HttpMethod.Get and early out in the if statement at line 1287, and save the extra null check.
```csharp
if (request.Content is null)
{
    if (request.Method == HttpMethod.Get)
    {
        // If this is a Get request and there is no content, then don't fill in the content as empty content gets rejected by some web services
        return;
    }
    ...
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (PaulHigin) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:373]: > I was running the tests locally and the numerous progress bars being rendered created a different problem (hitting a debug assert)

Funny, but I recently had the same assert issue with tests for PSGetV3 and also suppressed progress bar in tests.  It looks like that assert is incorrect and may affect other tests.  Can you create an issue and/or PR to remove the assert?

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:373]: Yeah, I'll take care of it in a separate PR

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: makes sense

File: cgmanifest.json
Comment (PaulHigin) [inline cgmanifest.json:507]: Looks like this new file somehow snuck into your PR.  I assume it should be removed?  Otherwise changes look good.

**Review by PaulHigin — APPROVED** [review]

File: cgmanifest.json
Comment (SteveL-MSFT) [reply cgmanifest.json:507]: The cgmanifest test complained and said I needed to update the list

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:373]: https://github.com/PowerShell/PowerShell/pull/16588

**Review by PaulHigin — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]
Ok, I think this makes sense.  There is no content but Get request expects content (and we know this because ContentType is non null) then we return empty content, and that is Ok with most servers.  It feels like there is no breaking change here since we were returning empty content before anyway.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: Are you sure we should ignore adding ContentHeaders below? I think no.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: Why do we ignore Header below? I think we need the headers.

Also if method is GET ContentType must be null https://stackoverflow.com/questions/5661596/do-i-need-a-content-type-header-for-http-get-requests
I'd expect this should be fixed on top of the method.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: Per your link, a `Get` without a body should NOT have a ContentType.  So in this case, if the user did not specify an explicit ContentType, that header isn't being added.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: I mean there can be other headers.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: I see what you are saying, but it doesn't apply here.  The loop below this is specifically for Content headers.  If the Content is null, then there are no headers.

**Comment by SteveL-MSFT** [comment]
Will pickup cgmanifest and static analysis failures from doc hyperlinks once those PRs are merged to master rather than having those fixes as part of this PR

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: Yes, but I did not find any differences in the code (in pwsh and .Net) for content headers and non-content headers.
So I'm worried about losing some headers like authorization.

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1286]: Other headers are handled elsewhere, this is specifically regarding Content headers

**Comment by iSazonov** [comment]
@SteveL-MSFT We get another fix for #15186 in the PR #16563. It seems it cover more scenarios but doesn't have tests. How we will resolve the situation?

**Comment by SteveL-MSFT** [comment]
@iSazonov unfortunately I didn't see that issue and the PR.  I think I'll combine that PR into this one to cover other `Get` cases.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:1285]: > CustomMethod.ToUpperInvariant() == "GET")

Nit: maybe use OrdinalIgnoreCase string comparison?

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@VeVrX Could you please review the PR?

**Comment by kmccl922** [comment]
You'll think I'm playing games with you'll i'm really getting the FBI to investigation on you'll sending all damn messages to my email please stop right now please

Sent from Yahoo Mail on Android 
 
  On Sat, Dec 18, 2021 at 2:14 AM, ***@***.***> wrote:   


@VeVrX Could you please review the PR?

**Review by adityapatwardhan — APPROVED** [review]

**Comment by TravisEz13** [comment]
@PowerShell/powershell-maintainers 
We only take security fixes and regression to LTS.  This is appropriate for 7.3.  This is not a regression from PowerShell 6.0 and newer.  There is some concern about stability.  Rejecting back-port for now.

Let's give this some bake time in 7.3 and reconsider.

**Comment by TravisEz13** [comment]
@PowerShell/powershell-committee This issue was introduced 2 LTS (during 7.0 preview) ago and doesn't meeting the servicing criteria that the maintainers understand they have the authority to approve.  Please clarify if this meets the backport criteria.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this.  Given this is a regression from 5.1 and PSCore6.x as well as real world impact based on the IETF RFC that we SHOULD not be sending content-length for empty payload, we recommend taking this to 7.2

**Pull request merged**
